lib/WebpackError.js

```diff
@@ -4,8 +4,81 @@
 */
 "use strict";
 
+const createHash = require("./util/createHash");
+
+const hash = Symbol();
+const meta = Symbol();
+
 module.exports = class WebpackError extends Error {
+	constructor(
+		message,
+		line = 0,
+		column = 0,
+		columnEnd = 0,
+		category = "error"
+	) {
+		// TODO: enable message assertion during construction in a commit to follow
+		// if (!message) {
+		// 	WebpackError.deprecate("message must be a non-null, non-empty value.");
+		// }
+		// TODO: enable message length limit at a commit to follow
+		// else if (message.length > 160) {
+		// 	WebpackError.deprecate("message cannot be longer than 80 characters.");
+		// }
+
+		super(message);
+
+		this[hash] = createHash("md4");
+		this[meta] = {};
+		this.category = category;
+		this.column = column;
+		this.columnEnd = columnEnd;
+		this.line = line;
+
+		// TODO: enable abstract protection at a commit to follow
+		// if (this.constructor === WebpackError) {
+		// 	const message = `A WebpackError cannot be directly constructed.`;
+		// 	WebpackError.deprecate(message);
+		// }
+	}
+
+	static deprecate(message) {
+		const err = new Error();
+		const stack = err.stack
+			.split("\n")
+			.slice(3, 4)
+			.join("\n");
+		// use process.stdout to assert the message will be displayed in
+		// environments where console has been proxied. this mimics node's
+		// util.deprecate method.
+		process.stdout.write(`DeprecationWarning: ${message}\n${stack}\n`);
```

Comment:
Why not using `process.emitWarning`? That's what is used by `util.deprecated`.
https://github.com/nodejs/node/blob/master/lib/internal/util.js#L60

Reply:
I found the `(node nnnn)` prefix to be superfluous in the log. It doesn't add any value, hence the util function didn't add any value.

Reply:
Yep that's important for `--no-deprecation` etc to work

Reply:
This flag sets `process.noDeprecation` which is handled by `util.deprecated`.

Reply:
https://nodejs.org/api/process.html#process_process_emitwarning_warning_type_code_ctor

> The following additional handling is implemented if the warning type is DeprecationWarning:
> *   If the --throw-deprecation command-line flag is used, the deprecation warning is thrown as an exception rather than being emitted as an event.
> *   If the --no-deprecation command-line flag is used, the deprecation warning is suppressed.
> *   If the --trace-deprecation command-line flag is used, the deprecation warning is printed to stderr along with the full stack trace.

Seems like there is some magic in emitWarning... ^^

Reply:
Since `util.deprecate` checks for `process.noDeprecation` I didn't look at `emitWarning` implementation. Then better use `emitWarning` I guess.

Reply:
OK are we on the same page that `process.emitWarning` is preferred here then?